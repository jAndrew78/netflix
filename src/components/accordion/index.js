import React, { useState } from 'react';


export default function Accordion({ children, ...restProps }) {
    <Container {...restProps}>
        <Inner>{children}</Inner>
    </Container>
}

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
    return <Title {...restProps}>{ children }</Title>;
}

Accordion.Frame = function AccordionFrame({ children, ...restProps }) {
    return <Frame {...restProps}>{ children }</Frame>;
}

Accordion.Item = function AccordionItem({ children, ...restProps }) {
    return <Item {...restProps}>{ children }</Item>;
}

Accordion.Body = function AccordionBody({ children, ...restProps }) {
    return <Body {...restProps}>{ children }</Body>;
}

