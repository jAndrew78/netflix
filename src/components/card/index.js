import React from 'react';
import { Container, Group, Title, SubTitle, Text, } from './styles/card';


export default function Card({ children, ...restProps }) {
    return <Container {...restProps}>{children}</Container>;
}

Card.Group = function CardGroup({ children, ...restProps }) {
    return <Group {...restProps}>{children}</Group>
}

Card.Title = function CardTitle({ children, ...restProps }) {
    return <Title {...restProps}>{children}</Title>
}

Card.SubTitle = function CardSubTitle({ children, ...restProps }) {
    return <SubTitle {...restProps}>{children}</SubTitle>
}

Card.Text = function CardText({ children, ...restProps }) {
    return <Text {...restProps}>{children}</Text>
}
