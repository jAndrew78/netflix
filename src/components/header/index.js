import React, { useState } from 'react';
import { Link as ReactRouterLink } from 'react-router-dom';

import { 
    Background, 
    Container, 
    Group, 
    Feature, 
    Logo, 
    ButtonLink, 
    Picture, 
    Profile,
    Dropdown,
    DropdownBreak,
    Break, 
    Text, 
    FeatureCallOut, 
    Link, 
    Search,
    SearchIcon,
    CloseIcon,
    SearchInput,
} from './styles/header';


export default function Header({ bg = true, children, ...restProps }) {
    return bg ? <Background {...restProps}>{children}</Background> : children;
}

Header.Frame = function HeaderFrame({ children, ...restProps }) {
    return <Container {...restProps}>{children}</Container>
}

Header.Group = function HeaderGroup({ children, ...restProps }) {
    return <Group {...restProps}>{children}</Group>
}

Header.Feature = function HeaderFeature({ children, ...restProps }) {
    return <Feature {...restProps}>{children}</Feature>
}

Header.Logo = function HeaderLogo({ to, ...restProps }) {
    return (
        <ReactRouterLink to={to}>
            <Logo {...restProps} />
        </ReactRouterLink>
    )
}

Header.ButtonLink = function HeaderButtonLink({ children, ...restProps }) {
    return <ButtonLink {...restProps}>{children}</ButtonLink>
}

Header.Picture = function HeaderPicture({ src, ...restProps }) {
    return <Picture {...restProps} src={`/images/users/${src}.png`} />
}

Header.Profile = function HeaderProfile({ children, ...restProps }) {
    return <Profile {...restProps}>{children}</Profile>
}

Header.Search = function HeaderSearch({ searchTerm, setSearchTerm, ...restProps }) {
    const [searchActive, setSearchActive] = useState(false);

    return (
        <Search {...restProps}>
            <SearchIcon onClick={() => setSearchActive(searchActive => !searchActive)} active={searchActive}>
                <img src="/images/icons/search.png" alt="Search" />
            </SearchIcon>
            <SearchInput
                value={searchTerm}
                onChange={({ target }) => setSearchTerm(target.value)}
                placeholder="Search films and series"
                active={searchActive}
            />
            <CloseIcon onClick={() => setSearchActive(searchActive => !searchActive)} active={searchActive}>
                <img src="/images/icons/close-slim.png" alt="Close" />
            </CloseIcon>
        </Search>
    ) 
}

Header.Dropdown = function HeaderDropdown({ children, ...restProps }) {
    return <Dropdown {...restProps}>{children}</Dropdown>
}

Header.DropdownBreak = function HeaderDropdownBreak({ children, ...restProps }) {
    return <DropdownBreak {...restProps}>{children}</DropdownBreak>
}

Header.Break = function HeaderBreak({ children, ...restProps }) {
    return <Break {...restProps}>{children}</Break>
}

Header.Text = function HeaderText({ children, ...restProps }) {
    return <Text {...restProps}>{children}</Text>
}

Header.FeatureCallOut = function HeaderFeatureCallOut({ children, ...restProps }) {
    return <FeatureCallOut {...restProps}>{children}</FeatureCallOut>
}

Header.TextLink = function HeaderTextLink({ children, ...restProps }) {
    return <Link {...restProps}>{children}</Link>
}

