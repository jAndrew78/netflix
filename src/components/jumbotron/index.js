import React from 'react';
import { Inner, Container } from './styles/jumbotron';


export default function Jumbotron({ children, direction = 'row', ...restProps }) {    // get children, pass direction, then allowing any other props to be passed after
    return(
        <Inner direction={direction}>
            <p>Hello Jumbo</p>
        </Inner>
    );
}

Jumbotron.Container = function JumbotronContainer({ children, ...restProps }) {
    return <Container { ...restProps }>{children}</Container>;
}
