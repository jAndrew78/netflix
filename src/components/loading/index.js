import React from 'react';
import { Container } from './styles/loading';


export default function Loading({ children, ...restProps }) {
    return <Container {...restProps}>{children}</Container>
}