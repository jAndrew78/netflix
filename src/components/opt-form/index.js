import React from 'react';
import { Container, Input, Button, Text } from './styles/opt-form';


export default function OptForm({ children, ...restProps }) {    // get children, then allow any other props to be passed after
    return <Container {...restProps}>{ children }</Container>;
}

OptForm.Input = function OptFormInput({ ...restProps }) {
    return <Input {...restProps} />;
}

OptForm.Button = function OptFormButton({ children, ...restProps }) {
    return <Button {...restProps}>{ children }</Button>;
}

OptForm.Text = function OptFormText({ children, ...restProps }) {
    return <Text {...restProps}>{ children }</Text>;
}

