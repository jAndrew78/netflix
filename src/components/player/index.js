import React, { useState, useContext, createContext } from 'react';
import ReactDOM from 'react-dom';

import { Container, Button, Overlay, Inner, Close } from './styles/player';


// have to use context with styled components to be able to pass state around
export const PlayerContext = createContext();

export default function Player ({ children, ...restProps }) {
    const [showPlayer, setShowPlayer] = useState(false);

    return (
        <PlayerContext.Provider value={{ showPlayer, setShowPlayer }}>
            <Container {...restProps}>{children}</Container>
        </PlayerContext.Provider>
    ) 
}

Player.Video = function PlayerVideo({ src, ...restProps }) {
    const { showPlayer, setShowPlayer} = useContext(PlayerContext);

    return showPlayer ? ReactDOM.createPortal(
        <Overlay {...restProps}> 
            <Inner>
                <video id="netflix-player" controls>
                    <source src={src} type="video/mp4" />
                </video>
                <Close onClick={() => setShowPlayer(false)}> 
                    <img src="/images/icons/close.png" alt="Close" />
                </Close>
            </Inner>
        </Overlay>,
        document.body
    ) : null;
}

Player.Button = function PlayerButton({ ...restProps }) {
    const { showPlayer, setShowPlayer } = useContext(PlayerContext);

    return (
        <Button 
            onClick={() => setShowPlayer(showPlayer => !showPlayer)}>
            {/* use below to remove showPlayer error above, but may cause an error in rare cases */}
            {/* onClick={() => setShowPlayer(!showPlayer)}> */}
            <span>&#9654; </span>
            Play
        </Button>
    );
}
