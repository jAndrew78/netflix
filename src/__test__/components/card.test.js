import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Card, Player } from '../../components';

const category = 'series';
const slideRows = [
    {
        title: 'Documentaries',
        data: [
            {
                genre: 'documentaries',
                maturity: '12',
                slug: 'man-on-wire',
                description: 'Man on Wire description',
                id: 'cdb04e38-4ea6-445b-9d03-6466ad034924',
                title: 'Man on Wire',
                docId: '2LJ4Y925uMsfDb1aggXV',
            },
        ],
    },
    {
        title: 'Feel Good',
        data: [
            {
                title: 'Juno',
                genre: 'feel-good',
                description: 'Juno description',
                maturity: '12',
                id: '1985cbe8-b6d7-442e-a955-d2d4f4175220',
                slug: 'juno',
                docId: 'KT7Qp8IylHQQs3fpHvxK',
            },
        ],
    },
    {
        title: 'children',
        data: [
            {
                title: 'Peppa Pig',
                genre: 'children',
                description: 'Peppa Pig description',
                maturity: '0',
                id: '506ef4fe-bc22-4da3-b8d9-35b9b541ab22',
                slug: 'peppa-pig',
                docId: 'tHhOMySyBDK70PzTfnh5',
            },
        ],
    },
];

describe('<Card />', () => {
    it('renders the <Card /> with populated data', () => {
        const { container, getByText } = render(
            <Card.Group>
                {slideRows.map((slideItem) => (
                    <Card key={`${category}-${slideItem.title.toLowerCase()}`}>
                        <Card.Title>{slideItem.title}</Card.Title>
                        
                        <Card.Entities>
                            {slideItem.data.map((item) => (
                                <Card.Item key={item.docId} item={item}>
                                    <Card.Image src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`} />

                                    <Card.Meta>
                                        <Card.SubTitle>{item.title}</Card.SubTitle>
                                        <Card.Text>{item.description}</Card.Text>
                                    </Card.Meta>
                                </Card.Item>
                            ))}
                        </Card.Entities>

                        <Card.Feature category={category}>
                            <Player>
                                <Player.Button />
                                <Player.Video src="/videos/bunny.mp4" />
                            </Player>
                        </Card.Feature>

                    </Card>
                ))}
            </Card.Group>
        );

        expect(getByText('Documentaries')).toBeTruthy();
        expect(getByText('Man on Wire')).toBeTruthy();
        expect(getByText('Man on Wire description')).toBeTruthy();
        
        expect(getByText('Feel Good')).toBeTruthy();
        expect(getByText('Juno')).toBeTruthy();
        expect(getByText('Juno description')).toBeTruthy();
        expect(container.firstChild).toBeTruthy();
    });

    it('renders the <Card /> and toggles the card feature', () => {
        const { container, queryByText, getByTestId, getByAltText } = render(
            <Card.Group>
                {slideRows.map((slideItem) => (
                    <Card key={`${category}-${slideItem.title.toLowerCase()}`}>
                        <Card.Title>{slideItem.title}</Card.Title>
                        
                        <Card.Entities>
                            {slideItem.data.map((item) => (
                                <Card.Item key={item.docId} item={item} data-testid={`${item.slug}-item-feature`}>
                                    <Card.Image src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`} />

                                    <Card.Meta>
                                        <Card.SubTitle>{item.title}</Card.SubTitle>
                                        <Card.Text>{item.description}</Card.Text>
                                    </Card.Meta>
                                </Card.Item>
                            ))}
                        </Card.Entities>

                        <Card.Feature category={category}>
                            <Player>
                                <Player.Button />
                                <Player.Video src="/videos/bunny.mp4" />
                            </Player>
                        </Card.Feature>

                    </Card>
                ))}
            </Card.Group>
        );

        expect(queryByText('MA-12')).toBeFalsy();
        fireEvent.click(getByTestId('juno-item-feature'));
        expect(queryByText('MA-12')).toBeTruthy();
        
        fireEvent.click(getByAltText('Close'));
        expect(queryByText('MA-12')).toBeFalsy();
        
        expect(queryByText('PG')).toBeFalsy();
        fireEvent.click(getByTestId('peppa-pig-item-feature'));
        expect(queryByText('PG')).toBeTruthy();
        
        fireEvent.click(getByAltText('Close'));
        expect(queryByText('PG')).toBeFalsy();
        expect(container.firstChild).toMatchSnapshot();
    });
});
