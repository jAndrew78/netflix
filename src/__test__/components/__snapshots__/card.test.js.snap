// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders the <Card /> and toggles the card feature 1`] = `
<div
  class="card__Group-s87lnr-2 Ugovi"
>
  <div
    class="card__Container-s87lnr-1 hCZzYq"
  >
    <p
      class="card__Title-s87lnr-0 cOUXPu"
    >
      Documentaries
    </p>
    <div
      class="card__Entities-s87lnr-11 hgmKeo"
    >
      <div
        class="card__Item-s87lnr-13 gfHnoL"
        data-testid="man-on-wire-item-feature"
      >
        <img
          class="card__Image-s87lnr-14 jAaaAr"
          src="/images/series/documentaries/man-on-wire/small.jpg"
        />
        <div
          class="card__Meta-s87lnr-12 hsksLA"
        >
          <p
            class="card__SubTitle-s87lnr-3 clCNLG"
          >
            Man on Wire
          </p>
          <p
            class="card__Text-s87lnr-4 kSFMJv"
          >
            Man on Wire description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card__Container-s87lnr-1 hCZzYq"
  >
    <p
      class="card__Title-s87lnr-0 cOUXPu"
    >
      Feel Good
    </p>
    <div
      class="card__Entities-s87lnr-11 hgmKeo"
    >
      <div
        class="card__Item-s87lnr-13 gfHnoL"
        data-testid="juno-item-feature"
      >
        <img
          class="card__Image-s87lnr-14 jAaaAr"
          src="/images/series/feel-good/juno/small.jpg"
        />
        <div
          class="card__Meta-s87lnr-12 hsksLA"
        >
          <p
            class="card__SubTitle-s87lnr-3 clCNLG"
          >
            Juno
          </p>
          <p
            class="card__Text-s87lnr-4 kSFMJv"
          >
            Juno description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card__Container-s87lnr-1 hCZzYq"
  >
    <p
      class="card__Title-s87lnr-0 cOUXPu"
    >
      children
    </p>
    <div
      class="card__Entities-s87lnr-11 hgmKeo"
    >
      <div
        class="card__Item-s87lnr-13 gfHnoL"
        data-testid="peppa-pig-item-feature"
      >
        <img
          class="card__Image-s87lnr-14 jAaaAr"
          src="/images/series/children/peppa-pig/small.jpg"
        />
        <div
          class="card__Meta-s87lnr-12 hsksLA"
        >
          <p
            class="card__SubTitle-s87lnr-3 clCNLG"
          >
            Peppa Pig
          </p>
          <p
            class="card__Text-s87lnr-4 kSFMJv"
          >
            Peppa Pig description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
